Update Function as a Service docs (#6970)
ballerina-platform/ballerina-dev-website: swan-lake/learn-the-platform/run-in-the-cloud/function-as-a-service/azure-functions.md

---
layout: ballerina-cloud-left-nav-pages-swanlake
title: Azure Functions
description: Learn how to write and deploy Azure Functions using Ballerina.
keywords: ballerina, programming language, serverless, cloud, azure, functions, cloud-native
permalink: /learn/run-ballerina-programs-in-the-cloud/function-as-a-service-with-ballerina/azure-functions/
active: azure-functions
intro: The Azure Functions extension provides the functionality to expose a Ballerina function as a serverless function in the Azure Functions platform.
---

> Azure Functions is an event-driven, serverless computing platform. Azure Functions can be written from Ballerina using the listeners and services provided by Azure Functions package. This example demonstrates using an HTTP trigger to invoke the function, and an HTTP output binding to respond to the caller with a message based on the query parameter sent from the request.

Author: Removed.

## Set up the prerequisites

Follow the instructions in the sections below to set up the prerequisites.

### Set up Azure Functions 

Follow the steps below to set up the Azure Functions.

1. Install the <a href="https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest" target="_blank">Azure CLI</a>.
2. Install and configure the [Functions Core Tools](https://learn.microsoft.com/en-us/azure/azure-functions/functions-run-local?tabs=v4%2Clinux%2Ccsharp%2Cportal%2Cbash#install-the-azure-functions-core-tools).

### Create the app

Follow the steps below to create the Azure function app.

1. Execute the `az login` command on the CLI to log in to the Azure CLI.
2. Create an <a href="https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-function-app-portal" target="_blank">Azure Function app</a> with the default option of creating a resource group automatically and the requirements below.
>**Note:** Make sure to remember the function application name and storage account name as they will be required in the code samples.

>       - Runtime stack - `Java 11`

Reviewer: These bullet points are not properly displayed in GH view file mode. 

<img width="1058" alt="image" src="https://github.com/ballerina-platform/ballerina-dev-website/assets/16300038/8537d496-b4b9-4782-a63c-4dc3687bcc00">

Author: Fixed.

      - Hosting operating system - `Windows` (Currently, Linux is not supported in Azure by default for custom handlers.)

## Create the function

You can write Azure functions that use different triggers to invoke the function, and output bindings to generate the response based on your use case. 

>**Info:** For examples of creating Azure functions, see [Examples](#examples).

## Build the function

The Azure Functions functionality is implemented as a compiler extension. Therefore, artifact generation happens automatically when you build a Ballerina module by executing the command below. 

```
$ bal build --cloud="azure_functions"
```

## Deploy the function

To deploy the function, execute the command, which you get in the CLI output logs after you [build the function](#build-the-function). For examples, see [Examples](#examples).

>**Note:** When you are deploying, make sure to replace the `<function_app_name>` placeholder with the app name of the [created function](#create-the-function).

> ## Azure Functions native (experimental)

Reviewer: Is this title complete or meaningful? Should it be `Azure Functions in native approach (experimental)` or similar? @xlight05

Reviewer 2: How about Azure Functions as a native executable?

Author: Updated.

> ## Azure Functions native (experimental)

Reviewer 2: Remove (experimental) part as graalvm is not experimental anymore.

Author: Done.

You can use Azure Functions in the [native approach of Ballerina](/learn/build-the-executable-in-a-container/).

> ### Set up the prerequisites for the native approach

Reviewer: Do we need to repeat `native approach` part in each subtitle?

Reviewer 2: +1 to remove

Author: Removed.

1. Set up the [general prerequisites](#set-up-the-prerequisites).
2. Install and configure [Docker](https://www.docker.com/) in your machine.
3. Install and configure [GraalVM](https://www.graalvm.org/) in your machine.

### Create an app for the native approach

Create an <a href="https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-function-app-portal" target="_blank">Azure Function app</a> with the given resource group with the following requirements.
   - Runtime stack - `Java 11`
   - Hosting operating system - `Linux` (As of now, Ballerina Azure functions native support is only available for Linux.)

### Create the function for the native approach

An Azure Functions package supports the two build options below.

>**Info:** Both build options for the non-native JVM-based approach will behave the same since the JVM is platform-independent.

- `cloud="azure_functions"` - Builds the native executable to be compatible with the Azure Functions cloud.
- `cloud="azure_functions_local"` - Builds the native executable to be compatible with your machine for development purposes.

> You can use these build options along with the native build option (i.e., `--native`) while building the package as shown below. 

Reviewer: ```suggestion
You can use these build options along with the native build option (i.e., `--native`) while building the package, as shown below. 
```

> You can use these build options along with the native build option (i.e., `--native`) while building the package, as shown below. 

Reviewer 2: ```suggestion
You can use these build options along with the graalvm build option (`--graalvm`) while building the package, as shown below. 
```

#### Build locally 

The example command below will build the package and run it locally.

> >**Info:** This will use the GraalVM you installed on your machine to build the native image and make the generated executable compatible with your machine.

Reviewer: Move info block after the command.

```bash

> $ bal build --cloud="azure_functions_local" --native

Reviewer 2: ```suggestion
$ bal build --cloud="azure_functions_local" --graalvm
```

```

#### Build for the cloud

The example command below will build the package for the Azure Functions cloud. 

> >**Info:** This will perform the compilation inside the Docker image to make it compatible with the Azure Functions cloud environment.

Reviewer: ditto

Author: Done.

### Set up Ballerina

Install the [latest Ballerina distribution](https://ballerina.io/downloads/).

### Set up Azure Functions 

Follow the steps below to set up the Azure Functions.

1. Install the <a href="https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest" target="_blank">Azure CLI</a>.
2. Install and configure [Functions Core Tools](https://learn.microsoft.com/en-us/azure/azure-functions/functions-run-local?tabs=v4%2Clinux%2Ccsharp%2Cportal%2Cbash#install-the-azure-functions-core-tools).

## Create the app

Follow the steps below to create the Azure function app.

1. Log in to the Azure CLI by executing the `az login` command on CLI.
2. Create an <a href="https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-function-app-portal" target="_blank">Azure Function app</a> with the given resource group with the following requirements.
>**Note:** Make sure to remember the function application name and storage account name as they will be required in the code samples.
   - Runtime stack - `Java 11`
   - Hosting operating system - `Windows` (Currently, Linux is not supported in Azure by default for custom handlers.)

> ## Write the function

Reviewer 2: Shall we rename the title to something like sample code? as we are only showing a generic example here?

Reviewer 2: This is applicable for aws lambda page as well

Author: The title `Write the function` connects well with the other topics in the flow of the use case.

Author: @anupama-pathirage - WDYT?

Author: Kept the title and added an intro sentence.

The following Ballerina code gives an example of using an HTTP trigger to invoke the function, and an HTTP output binding to respond to the caller with a message based on the query parameter sent from the request. 

```ballerina
import ballerinax/azure_functions as af;

service / on new af:HttpListener() {
    resource function get hello(string name) returns string {
        return "Hello, " + name + "!";
    }
}
```

## Build the function

The Azure Functions functionality is implemented as a compiler extension. Thus, artifact generation happens automatically when you build a Ballerina module. 

## Deploy the function

> The created function app name should be provided to the placeholders shown in the above-generated usage instructions from the compiler. 

Author: Updated.

## Invoke the function

> The deployed Azure Function can be tested by invoking it using an HTTP client such as cURL.

Reviewer 2: Same as AWS lambda, each function can have different triggers. Ex - if the trigger is queue, we can't use curl to invoke it. Therefore, we need to decide if we are keeping this section.

Author: Removed as invoking will be demonstrated in the BBEs.

## Azure Functions native (experimental)

You can use Azure Functions in the native approach of Ballerina, which is currently an experimental feature.

>**Info:** For more information on the native approach, see [Build a native executable](/learn/build-a-native-executable/#pack-the-native-executable-in-a-container). 

### Set up the prerequisites for the native approach

Follow the steps below to set up the prerequisites that are specifically required to follow the native approach.

1. Set up the [general prerequisites](#set-up-the-prerequisites).

> 2. Log in to the Azure CLI by executing the `az login` command on CLI.

Author: Removed.
